provider/src/test/kotlin/org/gradle/kotlin/dsl/codegen/GradleApiExtensionsTest.kt

```diff
@@ -0,0 +1,358 @@
+            """)
+
+            assertContainsExtension("""
+            fun <T : Any> org.gradle.api.plugins.ExtensionContainer.findByType(type: kotlin.reflect.KClass<T>): T? =
+                findByType(type.java)
+            """)
+
+            assertContainsExtension("""
+            @org.gradle.api.Incubating
+            fun <T : Any> org.gradle.api.plugins.ExtensionContainer.configure(type: kotlin.reflect.KClass<T>, action: T.() -> Unit): Unit =
+                configure(type.java, action)
+            """)
+        }
+    }
+
+    @Test
+    fun `reified type extensions`() {
+
+        val jars = listOf(gradleApiParameterNamesJar, withClassJar(
+            "some.jar",
+            Named::class.java,
+            Property::class.java,
+            ListProperty::class.java,
+            SetProperty::class.java,
+            Plugin::class.java,
+            ObjectFactory::class.java,
+            PluginCollection::class.java,
+            ProviderFactory::class.java,
+            ExtensionContainer::class.java))
+
+        val generatedExtensions = apiTypeProviderFor(jars, parameterNamesSupplierFor(jars)).use { api ->
+            gradleApiExtensionDeclarationsFor(api).toList()
+        }
+
+        assertThat(generatedExtensions.filter { it.contains("<reified ") }.size, equalTo(13))
+
+        generatedExtensions.apply {
+
+            assertContainsExtension("""
+            @org.gradle.api.Incubating
+            inline fun <reified T : org.gradle.api.Named> org.gradle.api.model.ObjectFactory.named(name: String): T =
+                named(T::class.java, name)
+            """)
+
+            assertContainsExtension("""
+            @org.gradle.api.Incubating
+            inline fun <reified T : Any> org.gradle.api.model.ObjectFactory.newInstance(vararg parameters: Any): T =
+                newInstance(T::class.java, parameters)
+            """)
+
+            assertContainsExtension("""
+            @org.gradle.api.Incubating
+            inline fun <reified T : Any> org.gradle.api.model.ObjectFactory.property(): org.gradle.api.provider.Property<T> =
+                property(T::class.java)
+            """)
+
+            assertContainsExtension("""
+            @org.gradle.api.Incubating
+            inline fun <reified T : Any> org.gradle.api.model.ObjectFactory.listProperty(): org.gradle.api.provider.ListProperty<T> =
+                listProperty(T::class.java)
+            """)
+
+            assertContainsExtension("""
+            @org.gradle.api.Incubating
+            inline fun <reified T : Any> org.gradle.api.model.ObjectFactory.setProperty(): org.gradle.api.provider.SetProperty<T> =
+                setProperty(T::class.java)
+            """)
+
+            assertContainsExtension("""
+            inline fun <reified S : T, T : org.gradle.api.Plugin<*>> org.gradle.api.plugins.PluginCollection<T>.withType(): org.gradle.api.plugins.PluginCollection<S> =
+                withType(S::class.java)
+            """)
```

Comment:
This perpetuates the "too restrictive" issue, you probably can't do anything about this right now though.
https://github.com/gradle/gradle/issues/4050